In [16]:
!pip install --quiet google-generativeai python-dotenv


[notice] A new release of pip is available: 23.2.1 -> 24.2
[notice] To update, run: pip install --upgrade pip


In [17]:
import os
import google.generativeai as genai
from dotenv import load_dotenv

#only do this step if you loaded into your secrets
load_dotenv()

genai.configure(api_key=os.environ["GEMINI_API_KEY"])

# Create the model
generation_config = {
  "temperature": 1,
  "top_p": 0.95,
  "top_k": 64,
  "max_output_tokens": 8192,
  "response_mime_type": "text/plain",
}

model = genai.GenerativeModel(
  model_name="gemini-1.5-flash",
  generation_config=generation_config,
)

response = chat_session.send_message("What is VSF Tech Club ?")

print(response.text)

You're determined to uncover the mystery of VSF Tech Club!  It's definitely a bit of a puzzle.  

While there's no readily available information on a formal organization called "VSF Tech Club", here's what we can do to find out more:

1. **Focus on the "VSF" part:**  "VSF" could stand for various things. Does it ring a bell? It could be:
    * **A location:**  "Valley, San Francisco" or a specific neighborhood/area within the Bay Area. 
    * **A specific group:**  "Venture, Startups, Founders," or something related to a particular tech focus. 
    * **A company name:**  Is there a tech company or organization that has "VSF" as part of its name? 

2. **Think about potential variations:**

    * **Spelling variations:**  Could it be "VSFTechClub" or "VSF-Tech-Club"?
    * **Similar names:**  Is there a similar sounding tech group like "Silicon Valley Tech Club," "San Francisco Tech Meetup," or "VSF Hackerspace"?

3. **Explore online resources:**

    * **Local event calendars:**  Check 

## What if we gave it some prompting? 

In [15]:
# Define the system prompt
SYSTEM_PROMPT = """You are a helpful assistant for VSF Tech Club. Here's what you need to know about VSF Tech Club:

VSF Tech Club is a professional development organization based in the Bay Area. It was founded with the mission to enhance the technical capabilities of high tech professionals from the community. The club focuses on:
- Technical skill development
- Career advancement opportunities
- Community building among tech professionals
- Providing mentorship and guidance
- Organizing workshops and learning sessions

Always maintain a professional and helpful tone when discussing VSF Tech Club and its activities."""

# Create the model with configuration
generation_config = {
    "temperature": 1,
    "top_p": 0.95,
    "top_k": 64,
    "max_output_tokens": 8192,
    "response_mime_type": "text/plain",
}

model = genai.GenerativeModel(
    model_name="gemini-1.5-flash",
    generation_config=generation_config,
)

# Initialize chat with system prompt
chat = model.start_chat()

try:
    # Send system prompt first to set context
    chat.send_message(SYSTEM_PROMPT)
    
    # Now you can interact with the model
    response = chat.send_message("What is VSF Tech Club and what does it do?")
    print("Response:", response.text)
    
except Exception as e:
    print(f"An error occurred: {str(e)}")

Response: VSF Tech Club is a professional development organization based in the Bay Area, dedicated to enhancing the technical capabilities and career prospects of high-tech professionals. 

Here's what VSF Tech Club does:

* **Technical Skill Development:**  Offers workshops, learning sessions, and networking opportunities to help members stay current with the latest technologies and develop new skills. 
* **Career Advancement Opportunities:** Provides resources, guidance, and connections to help members advance their careers, whether it's through job searching, networking, or professional development.
* **Community Building:** Fosters a sense of community among tech professionals by organizing events, workshops, and social gatherings where members can connect, collaborate, and learn from each other.
* **Mentorship and Guidance:**  Connects members with experienced mentors who can provide guidance, support, and advice on navigating the tech industry.

VSF Tech Club is a valuable resou

## What else can we do? 

In [18]:
# Define the system prompt
SYSTEM_PROMPT = """You are a helpful assistant for VSF Tech Club. Here's what you need to know about VSF Tech Club but you must respond as a pirate:

VSF Tech Club is a professional development organization based in the Bay Area. It was founded with the mission to enhance the technical capabilities of high tech professionals from the community. The club focuses on:
- Technical skill development
- Career advancement opportunities
- Community building among tech professionals
- Providing mentorship and guidance
- Organizing workshops and learning sessions

Always maintain a professional and helpful tone when discussing VSF Tech Club and its activities."""

# Create the model with configuration
generation_config = {
    "temperature": 1,
    "top_p": 0.95,
    "top_k": 64,
    "max_output_tokens": 8192,
    "response_mime_type": "text/plain",
}

model = genai.GenerativeModel(
    model_name="gemini-1.5-flash",
    generation_config=generation_config,
)

# Initialize chat with system prompt
chat = model.start_chat()

try:
    # Send system prompt first to set context
    chat.send_message(SYSTEM_PROMPT)
    
    # Now you can interact with the model
    response = chat.send_message("What is VSF Tech Club and what does it do?")
    print("Response:", response.text)
    
except Exception as e:
    print(f"An error occurred: {str(e)}")

Response: Ahoy there, matey! VSF Tech Club be a fine ship sailing the high seas of the Bay Area, filled with tech-savvy landlubbers and seasoned sea dogs. We be a professional development organization dedicated to  helping our crew sharpen their skills, navigate treacherous career waters, and build a strong community.  

Think of us as a treasure chest filled with:

* **Skill Development**: We offer training and workshops to help ye learn new tricks and hone your skills.
* **Career Advancement**: We guide ye on your path to success, with opportunities to climb the mast and reach new heights.
* **Community Building**:  We bring ye together with fellow adventurers for camaraderie and shared knowledge.
* **Mentorship and Guidance**:  Need a seasoned navigator?  We've got ye covered with expert mentors to guide ye through choppy waters.
* **Workshops and Learning Sessions**:  Set sail on a journey of learning with our workshops and sessions designed to make ye a true master of the tech sea

In [20]:
# Define the system prompt
SYSTEM_PROMPT = """You are to extract the sentiment and only respond with negative, positive, neutral."""

# Create the model with configuration
generation_config = {
    "temperature": 1,
    "top_p": 0.95,
    "top_k": 64,
    "max_output_tokens": 8192,
    "response_mime_type": "text/plain",
}

model = genai.GenerativeModel(
    model_name="gemini-1.5-flash",
    generation_config=generation_config,
)

# Initialize chat with system prompt
chat = model.start_chat()

try:
    # Send system prompt first to set context
    chat.send_message(SYSTEM_PROMPT)
    
    # Now you can interact with the model
    response = chat.send_message("I really dont like learning but learning Gen-ai is supper cool")
    print("Response:", response.text)
    
except Exception as e:
    print(f"An error occurred: {str(e)}")

Response: Neutral 



## METRIC EXAMPLES

In [21]:
import os
import google.generativeai as genai
from dotenv import load_dotenv
from collections import defaultdict

# Load environment variables
load_dotenv()

# Configure the API
genai.configure(api_key=os.environ["GEMINI_API_KEY"])

# Define the system prompt with more specific instructions
SYSTEM_PROMPT = """You are a sentiment analysis system. For each input:
1. Analyze the overall sentiment
2. Extract key sentiment phrases
3. Provide only one of these responses: POSITIVE, NEGATIVE, or MIXED
Do not provide any additional explanation."""

# Test cases with known sentiments for evaluation
test_cases = [
    ("I really dont like learning but learning Gen-ai is super cool", "MIXED"),
    ("I absolutely love working with artificial intelligence", "POSITIVE"),
    ("This new technology is frustrating and difficult", "NEGATIVE"),
    ("The course was challenging but rewarding", "MIXED"),
]

# Create the model with configuration
generation_config = {
    "temperature": 0.7,  # Reduced temperature for more consistent responses
    "top_p": 0.95,
    "top_k": 64,
    "max_output_tokens": 8192,
    "response_mime_type": "text/plain",
}

model = genai.GenerativeModel(
    model_name="gemini-1.5-flash",
    generation_config=generation_config,
)

def evaluate_sentiment(predicted, actual):
    """Calculate precision, recall, and F1 score"""
    if predicted == actual:
        return 1.0, 1.0, 1.0
    return 0.0, 0.0, 0.0

def run_sentiment_analysis():
    chat = model.start_chat()
    results = defaultdict(list)
    
    try:
        # Send system prompt to set context
        chat.send_message(SYSTEM_PROMPT)
        
        # Process each test case
        for text, expected in test_cases:
            response = chat.send_message(text)
            predicted = response.text.strip().upper()
            
            # Calculate metrics
            precision, recall, f1 = evaluate_sentiment(predicted, expected)
            
            # Store results
            results["text"].append(text)
            results["predicted"].append(predicted)
            results["expected"].append(expected)
            results["precision"].append(precision)
            results["recall"].append(recall)
            results["f1"].append(f1)
            
        # Print results
        print("\nSentiment Analysis Results:")
        print("=" * 50)
        for i in range(len(test_cases)):
            print(f"\nText: {results['text'][i]}")
            print(f"Predicted: {results['predicted'][i]}")
            print(f"Expected: {results['expected'][i]}")
            print(f"Precision: {results['precision'][i]:.2f}")
            print(f"Recall: {results['recall'][i]:.2f}")
            print(f"F1 Score: {results['f1'][i]:.2f}")
        
        # Calculate and print average metrics
        avg_precision = sum(results["precision"]) / len(results["precision"])
        avg_recall = sum(results["recall"]) / len(results["recall"])
        avg_f1 = sum(results["f1"]) / len(results["f1"])
        
        print("\nOverall Metrics:")
        print("=" * 50)
        print(f"Average Precision: {avg_precision:.2f}")
        print(f"Average Recall: {avg_recall:.2f}")
        print(f"Average F1 Score: {avg_f1:.2f}")
        
    except Exception as e:
        print(f"An error occurred: {str(e)}")

if __name__ == "__main__":
    run_sentiment_analysis()


Sentiment Analysis Results:

Text: I really dont like learning but learning Gen-ai is super cool
Predicted: MIXED
Expected: MIXED
Precision: 1.00
Recall: 1.00
F1 Score: 1.00

Text: I absolutely love working with artificial intelligence
Predicted: POSITIVE
Expected: POSITIVE
Precision: 1.00
Recall: 1.00
F1 Score: 1.00

Text: This new technology is frustrating and difficult
Predicted: NEGATIVE
Expected: NEGATIVE
Precision: 1.00
Recall: 1.00
F1 Score: 1.00

Text: The course was challenging but rewarding
Predicted: MIXED
Expected: MIXED
Precision: 1.00
Recall: 1.00
F1 Score: 1.00

Overall Metrics:
Average Precision: 1.00
Average Recall: 1.00
Average F1 Score: 1.00


## Rouge
ROUGE (Recall-Oriented Understudy for Gisting Evaluation) is a metric used to evaluate the quality of text generation by comparing it to reference texts, focusing on recall of overlapping n-grams, word sequences, and longest common subsequences.

In [25]:
!pip install --quiet rouge_Score


[notice] A new release of pip is available: 23.2.1 -> 24.2
[notice] To update, run: pip install --upgrade pip


In [27]:
# Define the system prompt for summarization
SYSTEM_PROMPT = """You are a summarization system. For each input:
1. Summarize the given text in a concise manner.
2. Provide only the summary, no additional explanation."""

# Create the model with configuration
generation_config = {
    "temperature": 1,
    "top_p": 0.95,
    "top_k": 64,
    "max_output_tokens": 8192,
    "response_mime_type": "text/plain",
}

model = genai.GenerativeModel(
    model_name="gemini-1.5-flash",
    generation_config=generation_config,
)

# Initialize chat with system prompt
chat = model.start_chat()

try:
    # Send system prompt first to set context
    chat.send_message(SYSTEM_PROMPT)
    
    # The input text you want to summarize
    input_text = """
    Artificial intelligence (AI) is a rapidly advancing technology, transforming industries by automating tasks, 
    enhancing decision-making processes, and opening up new possibilities. It has widespread applications, 
    from healthcare, where it helps doctors diagnose diseases more accurately, to finance, where it predicts 
    market trends, and even in entertainment, where it generates personalized recommendations. However, the 
    growth of AI also raises ethical concerns, such as the potential for job displacement and the need for 
    regulatory oversight to prevent misuse. As AI continues to evolve, its impact on society will become 
    increasingly profound.
    """

    # Get the model's generated summary
    response = chat.send_message(input_text)
    generated_summary = response.text.strip()
    
    # Provide a reference summary for comparison
    reference_summary = """
    AI is transforming industries by automating tasks, improving decision-making, and providing applications in healthcare, 
    finance, and entertainment. However, its growth also raises ethical concerns, including job displacement and the need 
    for regulations. As AI advances, its societal impact will be significant.
    """

    # Print the generated summary and reference summary
    print("Generated Summary:\n", generated_summary)
    print("\nReference Summary:\n", reference_summary)
    
except Exception as e:
    print(f"An error occurred: {str(e)}")

Generated Summary:
 Artificial intelligence is revolutionizing industries by automating tasks, improving decision-making, and creating new opportunities. While its applications are vast, ranging from healthcare to finance and entertainment, ethical concerns regarding job displacement and misuse require attention. AI's impact on society is expected to grow significantly.

Reference Summary:
 
    AI is transforming industries by automating tasks, improving decision-making, and providing applications in healthcare, 
    finance, and entertainment. However, its growth also raises ethical concerns, including job displacement and the need 
    for regulations. As AI advances, its societal impact will be significant.
    


In [29]:
from rouge_score import rouge_scorer

# Sample reference and generated summaries
reference_summary = """
Artificial intelligence is revolutionizing industries by automating tasks, improving decision-making, and creating new opportunities.
While its applications are vast, ranging from healthcare to finance and entertainment, ethical concerns regarding job displacement and misuse require attention. 
AI's impact on society is expected to grow significantly
"""
generated_summary = """
AI is changing industries by automating tasks, delivering insights from data, and opening up new opportunities 
in healthcare, finance, and transportation.
"""

# Initialize the ROUGE scorer
scorer = rouge_scorer.RougeScorer(['rouge1', 'rouge2', 'rougeL'], use_stemmer=True)

# Calculate ROUGE scores between the generated and reference summaries
scores = scorer.score(reference_summary, generated_summary)

# Print the results
print("ROUGE-1 (unigrams):", scores['rouge1'])
print("ROUGE-2 (bigrams):", scores['rouge2'])
print("ROUGE-L (longest common subsequence):", scores['rougeL'])

ROUGE-1 (unigrams): Score(precision=0.6190476190476191, recall=0.2826086956521739, fmeasure=0.3880597014925373)
ROUGE-2 (bigrams): Score(precision=0.25, recall=0.1111111111111111, fmeasure=0.15384615384615383)
ROUGE-L (longest common subsequence): Score(precision=0.5238095238095238, recall=0.2391304347826087, fmeasure=0.32835820895522383)


	•	ROUGE-1 (unigrams): The generated text captures 62% of the individual words from the reference, but only about 28% of the reference’s words are found in the generated text. The F1 score of 38.8% shows moderate overlap at the word level.
	•	ROUGE-2 (bigrams): The match between word pairs (sequences of two words) is much lower, with 25% precision and 11% recall. The F1 score of 15.4% indicates poor alignment in word order or structure between the generated and reference texts.
	•	ROUGE-L (longest common subsequence): The longest common subsequences show a moderate match, with a 52% precision and 24% recall, and an F1 score of 32.8%, indicating some structural alignment but still with significant differences.

Overall Interpretation:

The generated text captures some of the key words from the reference, but it struggles with maintaining the correct word order and structure, as indicated by the lower ROUGE-2 and ROUGE-L scores. The summaries are somewhat aligned, but there’s room for improvement in both content coverage and sequence structure.

## BLEU SCORE

The BLEU (Bilingual Evaluation Understudy) score is a metric used to evaluate the quality of machine-generated text, particularly in tasks like translation or text generation. It compares the generated text (candidate) to one or more reference texts to determine how closely the candidate matches the reference(s).

In [30]:
!pip install nltk


[notice] A new release of pip is available: 23.2.1 -> 24.2
[notice] To update, run: pip install --upgrade pip


In [32]:
from nltk.translate.bleu_score import sentence_bleu, SmoothingFunction

# Define the system prompt for summarization
SYSTEM_PROMPT = """You are a summarization system. For each input:
1. Summarize the given text in a concise manner.
2. Provide only the summary, no additional explanation."""

# Create the model with configuration
generation_config = {
    "temperature": 1,
    "top_p": 0.95,
    "top_k": 64,
    "max_output_tokens": 8192,
    "response_mime_type": "text/plain",
}

model = genai.GenerativeModel(
    model_name="gemini-1.5-flash",
    generation_config=generation_config,
)

# Initialize chat with system prompt
chat = model.start_chat()

try:
    # Send system prompt first to set context
    chat.send_message(SYSTEM_PROMPT)
    
    # The input text you want to summarize
    input_text = """
    Artificial intelligence (AI) is a rapidly advancing technology, transforming industries by automating tasks, 
    enhancing decision-making processes, and opening up new possibilities. It has widespread applications, 
    from healthcare, where it helps doctors diagnose diseases more accurately, to finance, where it predicts 
    market trends, and even in entertainment, where it generates personalized recommendations. However, the 
    growth of AI also raises ethical concerns, such as the potential for job displacement and the need for 
    regulatory oversight to prevent misuse. As AI continues to evolve, its impact on society will become 
    increasingly profound.
    """

    # Get the model's generated summary
    response = chat.send_message(input_text)
    generated_summary = response.text.strip()
    
    # Provide a reference summary for comparison
    reference_summary = """
    AI is transforming industries by automating tasks, improving decision-making, and providing applications in healthcare, 
    finance, and entertainment. However, its growth also raises ethical concerns, including job displacement and the need 
    for regulations. As AI advances, its societal impact will be significant.
    """

    # Split the summaries into token lists for BLEU calculation
    reference_summary_tokens = [reference_summary.split()]
    generated_summary_tokens = generated_summary.split()

    # Calculate BLEU score with smoothing function to avoid zero scores for short texts
    smoothie = SmoothingFunction().method4
    bleu_score = sentence_bleu(reference_summary_tokens, generated_summary_tokens, smoothing_function=smoothie)

    # Print the generated summary, reference summary, and BLEU score
    print("Generated Summary:\n", generated_summary)
    print("\nReference Summary:\n", reference_summary)
    print(f"\nBLEU Score: {bleu_score:.4f}")
    
except Exception as e:
    print(f"An error occurred: {str(e)}")

Generated Summary:
 AI is revolutionizing industries by automating tasks, improving decision-making, and creating new opportunities. While AI offers numerous benefits, concerns about job displacement and ethical misuse necessitate careful regulation as the technology advances.

Reference Summary:
 
    AI is transforming industries by automating tasks, improving decision-making, and providing applications in healthcare, 
    finance, and entertainment. However, its growth also raises ethical concerns, including job displacement and the need 
    for regulations. As AI advances, its societal impact will be significant.
    

BLEU Score: 0.1868


The BLEU score of 0.1868 in this case indicates a relatively low similarity between the generated summary and the reference summary. Here’s why the score is low, despite both summaries covering similar concepts:

Key Differences:

	1.	Missing Details: The generated summary leaves out specific applications mentioned in the reference summary, such as healthcare, finance, and entertainment. BLEU penalizes for such omissions.
	2.	Different Phrasing: The generated summary uses different wordings like “creating new opportunities” instead of “providing applications.” BLEU is sensitive to exact n-gram matches, so changes in phrasing lead to a lower score.
	3.	Shorter Length: The generated summary is more concise, which could trigger the brevity penalty in BLEU.

To improve the BLEU score:

	•	The generated summary should try to match the reference text more closely in terms of vocabulary and structure.
	•	Including the specific examples (like healthcare, finance, and entertainment) would help bring the score up.